src/scheduler/fifo.c

```diff
@@ -2240,29 +2266,44 @@ next_job(status *policy, server_info *sinfo, int flag)
 			 * start from the first queue of the next index
 			 */
 			if (queues_finished == queue_index_size) {
-							last_queue = 0;
-				last_index++;
+				last_queue = 0;
+				last_queue_index++;
 				i++;
-						}
+			}
 			queues_finished = 0;
-					}
-				}
-				else if (policy->by_queue) {
+		}
+	} else if (policy->by_queue) {
 		if (skip != SKIP_NON_NORMAL_JOBS) {
-			rjob = find_non_normal_job(sinfo->jobs);
-			if (rjob == NULL)
+			ind = find_non_normal_job(sinfo->jobs, last_job_index);
+			if (ind == -1) {
 				/* No more preempted jobs */
 				skip = SKIP_NON_NORMAL_JOBS;
+				last_job_index = 0;
+			} else {
+				rjob = sinfo->jobs[ind];
+				last_job_index = ind;
+			}
 		}
 		if (skip == SKIP_NON_NORMAL_JOBS) {
-					while (last_queue < sinfo->num_queues &&
-			    (rjob = find_runnable_resresv(sinfo->queues[last_queue]->jobs)) == NULL) {
-						last_queue++;
-					}
-				}
-	}
-				else	 { /* treat the entire system as one large queue */
-		rjob = find_runnable_resresv(sinfo->jobs);
+			while(last_queue < sinfo->num_queues &&
+			     ((ind = find_runnable_resresv(sinfo->queues[last_queue]->jobs, last_job_index)) == -1)) {
+				last_queue++;
+				last_job_index = 0;
+			}
+			if (ind != -1) {
```

Comment:
In practice this won't happen.  Once we reach the end of the last queue, we'll return -1.  It is good to be diligent though.  I added an extra check to make sure we didn't dereference a NULL pointer (because sinfo->queues[sinfo->num_queues] is the NULL sentinel)